pkg/tls/cred.go

```diff
@@ -0,0 +1,186 @@
+	chain := make([][]byte, len(crt.TrustChain)+1)
+	chain[0] = crt.Certificate.Raw
+	for i, c := range crt.TrustChain {
+		chain[len(crt.TrustChain)-i] = c.Raw
+	}
+	return chain
+}
+
+// EncodePEM emits a certificate and trust chain as a
+// series of PEM-encoded certificates from leaf to root.
+func (crt *Crt) EncodePEM() string {
+	buf := bytes.Buffer{}
+	encode(&buf, &pem.Block{Type: "CERTIFICATE", Bytes: crt.Certificate.Raw})
+
+	// Serialize certificates from leaf to root.
+	n := len(crt.TrustChain)
+	for i := n - 1; i >= 0; i-- {
+		encode(&buf, &pem.Block{Type: "CERTIFICATE", Bytes: crt.TrustChain[i].Raw})
+	}
+
+	return buf.String()
+}
+
+// EncodeCertificatePEM emits the Crt's leaf certificate as PEM-encoded text.
+func (crt *Crt) EncodeCertificatePEM() string {
+	return EncodeCertificatesPEM(crt.Certificate)
+}
+
+// EncodePrivateKeyPEM emits the private key as PEM-encoded text.
+func (cred *Cred) EncodePrivateKeyPEM() string {
+	b, err := x509.MarshalECPrivateKey(cred.PrivateKey)
+	if err != nil {
+		panic(fmt.Sprintf("Invalid elliptic curve: %s", err))
+	}
+
+	return string(pem.EncodeToMemory(&pem.Block{Type: "EC PRIVATE KEY", Bytes: b}))
+}
+
+// EncodePrivateKeyP8 encodes the provided key to the PKCS#8 binary form.
+func (cred *Cred) EncodePrivateKeyP8() ([]byte, error) {
+	return x509.MarshalPKCS8PrivateKey(cred.PrivateKey)
+}
+
+// certificateMatchesKey returns whether the key and certificate match.
+func certificateMatchesKey(c *x509.Certificate, k *ecdsa.PrivateKey) (ok bool) {
+	pub, ok := c.PublicKey.(*ecdsa.PublicKey)
+	ok = ok && pub.X.Cmp(k.X) == 0 && pub.Y.Cmp(k.Y) == 0
+	return
+}
+
+// SignCrt uses this Cred to sign a new certificate.
+//
+// This may fail if the Cred contains an end-entity certificate.
+func (cred *Cred) SignCrt(template *x509.Certificate) (*Crt, error) {
+	crtb, err := x509.CreateCertificate(
+		rand.Reader,
+		template,
+		cred.Crt.Certificate,
+		template.PublicKey,
+		cred.PrivateKey,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := x509.ParseCertificate(crtb)
+	if err != nil {
+		return nil, err
+	}
+
+	crt := Crt{
+		Certificate: c,
+		TrustChain:  append(cred.Crt.TrustChain, cred.Crt.Certificate),
+	}
+	return &crt, nil
+}
+
+// ReadPEMCreds reads PEM-encoded credentials from the named files.
+func ReadPEMCreds(keyPath, crtPath string) (*Cred, error) {
+	keyb, err := ioutil.ReadFile(keyPath)
+	if err != nil {
+		return nil, err
+	}
+	key, err := DecodePEMKey(string(keyb))
+	if err != nil {
+		return nil, err
+	}
+
+	crtb, err := ioutil.ReadFile(crtPath)
+	if err != nil {
+		return nil, err
+	}
+	crt, err := DecodePEMCrt(string(crtb))
+	if err != nil {
+		return nil, err
+	}
+	if !certificateMatchesKey(crt.Certificate, key) {
+		return nil, errors.New("tls: Public and private key do not match")
+	}
+	return &Cred{PrivateKey: key, Crt: *crt}, nil
+}
+
+// DecodePEMCrt decodes PEM-encoded certificates from leaf to root.
+func DecodePEMCrt(txt string) (*Crt, error) {
```

Comment:
Just noting that `ReadPEMCreds` and `DecodePEMCrt` are unused in this branch, but totally fine to keep them if they're intended for some future change.

Reply:
Yeah, they're used on my other branch...